scripts/export_environment.py

```diff
@@ -170,32 +160,45 @@ def resolve_env(env: Environment) -> Environment:
     >>> resolve_env({'x': '{y}', 'y': '42'})
     {'x': '42', 'y': '42'}
 
-    Literal curly braces need to be escaped:
+    Transitive references are supported:
 
-    >>> resolve_env({'x': '{{', 'y': '}}', 'z': '{{x}}'})
-    {'x': '{', 'y': '}', 'z': '{x}'}
+    >>> resolve_env({'x': '{y}', 'y': '{z}', 'z': '42'})
+    {'x': '42', 'y': '42', 'z': '42'}
 
-    Does not support transitive references:
+    Generative references formed during resolution are allowed provided that
+    curly braces are escaped:
 
-    >>> resolve_env({'x': '{y}', 'y': '{z}', 'z': '42'})
-    {'x': '{z}', 'y': '42', 'z': '42'}
+    >>> resolve_env({'x': '{{{y}}}', 'y': 'z', 'z': '42'})
+    {'x': '42', 'y': 'z', 'z': '42'}
 
-    Transitive resolution would yield
+    Compared to:
 
-    {'x': '42', 'y': '42', 'z': '42'}
+    >>> resolve_env({'x': '{{y}}', 'y': 'z', 'z': '42'})
+    {'x': 'z', 'y': 'z', 'z': '42'}
 
-    Does not support generative ones (references formed during resolution):
+    References to a missing key yields an empty string replacement:
 
-    >>> resolve_env({'x': '{y}}}', 'y': '{{z', 'z': '42'})
-    {'x': '{{z}', 'y': '{z', 'z': '42'}
+    >>> resolve_env({'x': 'y is {y}'})
+    {'x': 'y is '}
 
-    which could be
+    Unmatched curly braces are not allowed:
 
-    {'x': '42', 'y': '{z', 'z': '42'}
+    >>> resolve_env({'x': '{y'})
+    Traceback (most recent call last):
+    ...
+    ValueError: expected '}' before end of string
     """
-    # `{missing_var}` references in values should yield the empty string,
-    # consistent with $missing_var in the shell.
-    return {k: v.format_map(defaultdict(str, env)) for k, v in env.items()}
+    while True:
+        resolved_env = {}
+        for k, v in env.items():
+            try:
+                resolved_env[k] = v.format_map(env)
+            except KeyError as e:
+                resolved_env[k] = v.replace('{' + e.args[0] + '}', '')
```

Comment:
This smells a little because it is emulating format_map's behaviour, which is tends to be brittle or incomplete.

Consider `'{a!s}'.format_map(dict(b=1))`

What's wrong with the defaultdict approach?

Reply:
Ah, the ticket description had a recommended change for the function that omitted the `defaultdict` so I thought it was desired to remove it. https://github.com/DataBiosphere/azul/issues/1645#issue-583472613

I've reimplemented `defaultdict` in my current solution.

Reply:
I think I added the defaultdict after I created #1645.